File: src/RqCalc.WebApp/_Scripts/Models/GuildTalentBranchModel.js
function GuildTalentBranchModel(context, windowModel, branch) {

    var thisModel = this;

    thisModel.branch = branch;
    thisModel.windowModel = windowModel;
    //-----------------------------------------------------------------------------------------------------------------

    thisModel.talents = branch.talents.map(function (tal) {
        return new GuildTalentModel(context, thisModel, tal);
    });


    function getPrevBranch() {

        return windowModel.branches.takeWhile(function (b) {
            return b !== thisModel;
        }).lastOrDefault();
    }

    function getNextBranch() {

        return windowModel.branches.skipWhile(function (b) {
            return b !== thisModel;
        }).skip(1).firstOrDefault();
    }

    thisModel.isMaximized = function () {
        return thisModel.talents.sum(function (t) { return t.points(); }) === branch.maxPoints;
    }

    thisModel.hasActiveTalents = function () {
        return thisModel.talents.any(function (t) {
            return t.active();
        });
    }

    thisModel.isEditable = function () {

        var prevBranch = getPrevBranch();

        if (prevBranch == null || prevBranch.isMaximized()) {

            var nextBranch = getNextBranch();

            if (nextBranch == null || !nextBranch.hasActiveTalents()) {
                return true;
            }
        }

        return false;
    }

    //-----------------------------------------------------------------------------------------------------------------
    thisModel.setActiveGuildTalents = function (talents) {

        thisModel.talents.foreach(function (talentModel) {

            var points = talents.getValueByKeyOrDefault(talentModel.talent);

            talentModel.points(points == null ? 0 : points);
        });
    };

    //-----------------------------------------------------------------------------------------------------------------
    thisModel.getActiveGuildTalents = function () {

        return thisModel.talents.filter(function (talentModel) {
            return talentModel.active();
        }).map(function (talentModel) {
            return { key: talentModel.talent, value: talentModel.points() };
        });
    };
}

File: src/RqCalc.Wpf/Models/GuildTalentBranchModel.cs
using Framework.Core;
using Framework.Reactive;

using RqCalc.Domain.GuildTalent;
using RqCalc.Wpf.Models.Window.Dialog;

namespace RqCalc.Wpf.Models;

public class GuildTalentBranchModel : NotifyModelBase
{
    public readonly GuildTalentsWindowModel WindowModel;


    public GuildTalentBranchModel(GuildTalentsWindowModel windowModel, IGuildTalentBranch talentBranch)

    {
        this.WindowModel = windowModel ?? throw new ArgumentNullException(nameof(windowModel));
        this.GuildTalentBranch = talentBranch ?? throw new ArgumentNullException(nameof(talentBranch));

        this.Talents = talentBranch.Talents.ToObservableCollection(tal => new GuildTalentModel(this, tal));
    }

    public GuildTalentBranchModel PrevBranch => this.WindowModel.Branches.TakeWhile(b => b != this).LastOrDefault();

    public GuildTalentBranchModel NextBranch => this.WindowModel.Branches.SkipWhile(b => b != this).Skip(1).FirstOrDefault();

    public bool Maximized => this.Talents.Sum(t => t.Points) == this.GuildTalentBranch.MaxPoints;

    public bool Editable => (this.PrevBranch?.Maximized ?? true) && !(this.NextBranch?.HasActiveTalents ?? false);

    public bool HasActiveTalents => this.Talents.Any(t => t.Active);

    public IReadOnlyDictionary<IGuildTalent, int> ActiveTalents
    {
        get
        {
            var request = from model in this.Talents

                where model.Points > 0

                select model.SelectedObject.ToKeyValuePair(model.Points);

            return request.ToDictionary();
        }
        set
        {
            foreach (var model in this.Talents)
            {
                model.Points = value.GetValueOrDefault(model.SelectedObject);
            }
        }
    }
        
    public ObservableCollection<GuildTalentModel> Talents
    {
        get => this.GetValue(v => v.Talents);
        private set => this.SetValue(v => v.Talents, value);
    }

    public IGuildTalentBranch GuildTalentBranch
    {
        get => this.GetValue(v => v.GuildTalentBranch);
        private set => this.SetValue(v => v.GuildTalentBranch, value);
    }
}